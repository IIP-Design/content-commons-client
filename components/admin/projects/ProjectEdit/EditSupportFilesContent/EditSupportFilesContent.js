/**
 *
 * EditSupportFilesContent
 *
 */
import React from 'react';
import { func, object, string } from 'prop-types';
import gql from 'graphql-tag';
import orderBy from 'lodash/orderBy';
import {
  Button, Form, Loader, Table
} from 'semantic-ui-react';

import ModalItem from 'components/modals/ModalItem/ModalItem';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
import EditSupportFileRow from 'components/admin/projects/ProjectEdit/EditSupportFileRow/EditSupportFileRow';
import graphqlDynamicHOC from 'components/admin/graphqlDynamicHOC/graphqlDynamicHOC';

import { capitalizeFirst } from 'lib/utils';

import './EditSupportFilesContent.scss';

/* eslint-disable react/prefer-stateless-function */
class EditSupportFilesContent extends React.PureComponent {
  getFileExtension = str => (
    str.slice( ( Math.max( 0, str.lastIndexOf( '.' ) ) || Infinity ) )
  )

  getFileExtensions = arr => {
    const allFileExtensions = arr.reduce( ( acc, curr ) => (
      acc.concat( this.getFileExtension( curr.filename ) )
    ), [] );
    const uniqueExtensions = [...new Set( allFileExtensions )];
    return uniqueExtensions;
  }

  handleCancelClose = () => {
    console.log( 'cancel' );
    this.props.closeEditModal();
  }

  handleAddFiles = () => {
    console.log( 'add files' );
    this.addFilesInputRef.click();
  }

  handleAddFilesRef = input => {
    this.addFilesInputRef = input;
  }

  renderRow = file => {
    const {
      orderByField,
      projectId,
      data: { project: { files } }
    } = this.props;

    return (
      <EditSupportFileRow
        key={ file.id }
        file={ file }
        orderByField={ orderByField }
        projectId={ projectId }
        fileExtensions={ this.getFileExtensions( files ) }
      />
    );
  }

  render() {
    if ( !this.props.data.project ) return null;

    const {
      fileType,
      data,
      data: { project: { files } }
    } = this.props;

    const isSrt = fileType === 'srt';

    if ( data.loading ) {
      return (
        <div
          className="edit-support-files-loader"
          style={ {
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
            minHeight: '200px'
          } }
        >
          <Loader
            active
            inline="centered"
            style={ { marginBottom: '1em' } }
          />
          <p>Loading support file(s)...</p>
        </div>
      );
    }

    if ( data.error ) return `Error! ${data.error.message}`;

    if ( !files || !files.length ) return null;

    const sortedFiles = orderBy( files, ['filetype'] );

    const headline = isSrt
      ? fileType.toUpperCase()
      : capitalizeFirst( fileType );

    return (
      <ModalItem
        className={ `edit-support-files ${fileType}` }
        headline={ `Edit ${headline} file${files.length > 1 ? 's' : ''} in this project` }
        textDirection="ltr"
      >
        <Form>
          <Table basic="very">
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell width={ isSrt ? 9 : 7 }>
                  { `${headline} File${files.length > 1 ? 's' : ''} Selected` }
                </Table.HeaderCell>
                <Table.HeaderCell width={ 4 }>
                  Language
                  <small className="msg--required"> *</small>
                </Table.HeaderCell>

                { !isSrt
                  && (
                    <Table.HeaderCell width={ 4 }>
                      Type/Use
                      <small className="msg--required"> *</small>
                    </Table.HeaderCell>
                  ) }

                <Table.HeaderCell />
              </Table.Row>
            </Table.Header>

            <Table.Body>
              { sortedFiles.map( this.renderRow ) }
            </Table.Body>
          </Table>

          <div className="btn-group">
            <Button
              className="cancel-close"
              content="Cancel"
              basic
              size="tiny"
              onClick={ this.handleCancelClose }
              type="button"
            />
            <Button
              className="add-files"
              content="Add Files"
              color="blue"
              basic
              size="tiny"
              onClick={ this.handleAddFiles }
              type="button"
            />
            <VisuallyHidden>
              { /* eslint-disable jsx-a11y/label-has-for */
                /* eslint-disable jsx-a11y/label-has-associated-control */ }
              <label htmlFor="upload-file--multiple">upload files</label>
              <input
                id="upload-file--multiple"
                ref={ this.handleAddFilesRef }
                type="file"
                multiple
                tabIndex={ -1 }
              />
            </VisuallyHidden>
          </div>
        </Form>
      </ModalItem>
    );
  }
}

EditSupportFilesContent.propTypes = {
  data: object.isRequired,
  projectId: string.isRequired,
  fileType: string,
  orderByField: string,
  closeEditModal: func
};

const SUPPORT_FILES_QUERY = props => (
  gql`
    query SupportFiles($id: ID!, $fileType: String!, $orderBy: SupportFileOrderByInput) {
      project: videoProject(id: $id) {
        files: supportFiles(
          where: { ${props.orderByField}: $fileType }
          orderBy: $orderBy
        ) {
          id
          filename
          filetype
          language {
            id
            displayName
          }
        }
      }
    }
  `
);

export default graphqlDynamicHOC( SUPPORT_FILES_QUERY, {
  options: props => ( {
    variables: {
      id: props.projectId,
      fileType: 'srt',
      orderBy: 'filename_ASC'
    }
  } )
} )( EditSupportFilesContent );

export { SUPPORT_FILES_QUERY };
