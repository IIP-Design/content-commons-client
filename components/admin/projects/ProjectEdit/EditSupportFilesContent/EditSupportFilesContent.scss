@import "../../../../../styles/colors.scss";

.edit-support-files {
  :focus:not(:focus-visible) {
    outline: none
  }

  .ui.form {
    font-size: 0.777778rem;
  }

  .ui.basic.buttons {
    border: 0;

    > .button {
      border-left: 0;
    }
  }

  .ui.button {
    // font-size: 0.777778rem;
    box-shadow: none !important;

    &.basic.replace,
    &.basic.delete {
      margin-left: 0.25em;
      margin-right: 0.25em;
      padding: 0;
    }

    &:focus,
    &.basic.buttons .button:focus,
    &.basic:focus {
      outline: 2px dotted $medium-grey;
      outline-offset: 2px;
    }

    &.cancel-close {
      border: 2px solid transparent;
      @media screen and (max-width: 899px) {
        border-color: $grey;
      }
      color: $grey !important;
      font-weight: normal;

      &:hover,
      &:active,
      &:focus {
        border-color: $grey;
        color: #fff !important;
        background-color: $grey !important;
      }
    }

    &.basic.add-files {
      border: 2px solid $matisse-blue;
      color: $matisse-blue !important;
      font-weight: bold;
      box-shadow: none !important;

      &:hover,
      &:active,
      &:focus {
        color: #fff !important;
        background-color: $matisse-blue !important;
      }
    }

    &.save {
      border: 2px solid $light-blue !important;
      color: #fff !important;
      background-color: $light-blue !important;
  
      &:hover,
      &:active,
      &:focus {
        border-color: $matisse-blue !important;
        background-color: $matisse-blue !important;
      }
    }

    &.disabled,
    &:disabled {
      cursor: not-allowed;
      pointer-events: auto !important;
    }
  }

  .ui.selection.dropdown {
    &.error {
      background: initial;

      > .menu {
        > .item {
          font-size: 1em;
          background: $chablis;

          &.selected {
            background-color: $linen;
          }
        }
      }

      &:focus {
        .menu {
          border-color: $cavern-pink;
        }
      }
    }
  }

  .ui.table {
    &.basic {
      tbody tr {
        border-bottom: 0;
      }
    }
  }

  .btn-group {
    padding-top: 1em;
    padding-bottom: 1em;
    text-align: right;
  }

  .msg--required {
    color: $red;
  }

  .hasEllipsis {
    cursor: default;
  }
}

.ui.dimmer.modals {
  .edit-support-files-loader {
    .ui.loader:before {
      border-color: rgba(0, 0, 0, 0.1);
    }
    .ui.loader:after {
      border-color: #767676 transparent transparent;
    }
  }
}
