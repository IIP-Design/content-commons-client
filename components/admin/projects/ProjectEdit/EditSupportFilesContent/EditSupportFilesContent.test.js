import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Icon, Loader } from 'semantic-ui-react';
import EditSupportFilesContent, { SUPPORT_FILES_QUERY } from './EditSupportFilesContent';

const props = {
  projectId: '123',
  fileType: 'other',
  closeEditModal: jest.fn(),
  orderByField: 'filetype_not',
};

const mocks = [
  {
    request: {
      query: SUPPORT_FILES_QUERY( props ),
      variables: {
        id: props.projectId,
        fileType: 'srt',
        orderBy: 'filename_ASC'
      }
    },
    result: {
      data: {
        project: {
          id: '123',
          files: [
            {
              id: 'euosiq',
              filename: 'file-1.jpg',
              filetype: 'jpg',
              use: {
                id: 'th89',
                name: 'Thumbnail/Cover Image'
              },
              language: { id: '8h8sof', displayName: 'English' }
            },
            {
              id: 'aewk81',
              filename: 'file-2.mp3',
              filetype: 'mp3',
              use: {
                id: 'me92',
                name: 'Memes'
              },
              language: { id: '9weijd', displayName: 'French' }
            }
          ]
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <EditSupportFilesContent { ...props } />
  </MockedProvider>
);

describe( '<EditSupportFilesContent />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
      />
    );

    expect( editSupportFilesContent.exists() ).toEqual( true );
    expect( editSupportFilesContent.contains( loader ) ).toEqual( true );
    expect( editSupportFilesContent.find( 'p' ).text() )
      .toEqual( 'Loading support file(s)...' );
    expect( toJSON( editSupportFilesContent ) ).toMatchSnapshot();
  } );

  it( 'renders error message & icon if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: SUPPORT_FILES_QUERY( props ),
          variables: {
            id: props.projectId,
            fileType: 'srt',
            orderBy: 'filename_ASC'
          }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks }>
        <EditSupportFilesContent { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const div = editSupportFilesContent
      .find( 'div.edit-support-files-content.error' );
    const icon = <Icon color="red" name="exclamation triangle" />;
    const span = <span>Loading error...</span>;

    expect( div.exists() ).toEqual( true );
    expect( editSupportFilesContent.contains( icon ) )
      .toEqual( true );
    expect( editSupportFilesContent.contains( span ) )
      .toEqual( true );
  } );

  it( 'renders `null` if project is `null`', async () => {
    const nullMocks = [
      {
        request: {
          query: SUPPORT_FILES_QUERY( props ),
          variables: {
            id: props.projectId,
            fileType: 'srt',
            orderBy: 'filename_ASC'
          }
        },
        result: {
          data: { project: null }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullMocks }>
        <EditSupportFilesContent { ...props } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );

    expect( editSupportFilesContent.html() ).toEqual( null );
  } );

  it( '`getFileExtension` returns a file\'s extension', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const inst = editSupportFilesContent.instance();
    const { files } = inst.props.data.project;

    files.forEach( ( file, i ) => {
      expect( inst.getFileExtension( file.filename ) )
        .toEqual( `.${files[i].filetype}` );
    } );
  } );

  it( '`getFileExtensions` returns an array of unique file extensions', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const inst = editSupportFilesContent.instance();
    const { files } = inst.props.data.project;
    const extensions = inst.getFileExtensions( files );

    expect( extensions ).toEqual( ['.jpg', '.mp3'] );
  } );

  it( 'clicking the Cancel button calls `closeEditModal`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const cancelCloseBtn = editSupportFilesContent.find( 'Button.cancel-close' );

    cancelCloseBtn.simulate( 'click' );

    expect( props.closeEditModal ).toHaveBeenCalled();
  } );

  it( 'assigns `addFilesInputRef` to the file `input` element', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const inst = editSupportFilesContent.instance();
    const input = editSupportFilesContent.find( 'input#upload-file--multiple' );

    expect( input.exists() ).toEqual( true );
    expect( inst.addFilesInputRef ).toBeTruthy();
  } );

  it( 'clicking Add Files button calls `handleAddFiles`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const editSupportFilesContent = wrapper.find( 'EditSupportFilesContent' );
    const addFilesBtn = wrapper.find( 'Button.add-files' );
    const inst = editSupportFilesContent.instance();
    const spy = jest.spyOn( inst, 'handleAddFiles' );

    inst.forceUpdate();
    addFilesBtn.simulate( 'click' );

    expect( spy ).toHaveBeenCalled();
  } );
} );
