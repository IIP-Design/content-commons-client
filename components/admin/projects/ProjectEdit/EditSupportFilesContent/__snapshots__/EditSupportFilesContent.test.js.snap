// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditSupportFilesContent /> renders initial loading state without crashing 1`] = `
<EditSupportFilesContent
  closeEditModal={[MockFunction]}
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "fileType": "srt",
        "id": "123",
        "orderBy": "filename_ASC",
      },
    }
  }
  fileType="other"
  orderByField="filetype_not"
  projectId="123"
>
  <div
    className="edit-support-files-loader"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      />
    </Loader>
    <p>
      Loading support file(s)...
    </p>
  </div>
</EditSupportFilesContent>
`;
